### PR TITLE
feat(MFA): Add SMTP resource and Email Factor

### PR DESCRIPTION
**Pull request by tcompiegne**
Fixes gravitee-io/issues#5166

### CODE REVIEW

File: gravitee-am-management-api/gravitee-am-management-api-services/gravitee-am-management-api-services-purge/src/main/java/io/gravitee/am/management/services/purge/TableName.java [hunk 1]
```diff
@@ -27,5 +27,6 @@
     request_objects,
     login_attempts,
     uma_permission_ticket,
-    auth_flow_ctx
+    auth_flow_ctx,
+    mfa_codes
```
Comment (tcompiegne) [inline gravitee-am-management-api/gravitee-am-management-api-services/gravitee-am-management-api-services-purge/src/main/java/io/gravitee/am/management/services/purge/TableName.java:31]: I think we can remove this one

File: gravitee-am-repository/gravitee-am-repository-jdbc/src/test/java/io/gravitee/am/repository/jdbc/common/AbstractJdbcTest.java [hunk 2]
```diff
@@ -100,6 +100,8 @@ public void truncateTables() throws Exception {
             tables.add("alert_triggers_alert_notifiers");
             tables.add("alert_triggers");
             tables.add("alert_notifiers");
+            tables.add("mfa_codes");
```
Comment (tcompiegne) [inline gravitee-am-repository/gravitee-am-repository-jdbc/src/test/java/io/gravitee/am/repository/jdbc/common/AbstractJdbcTest.java:103]: To remove

File: gravitee-am-repository/gravitee-am-repository-jdbc/src/test/java/io/gravitee/am/repository/jdbc/common/JdbcRepositoriesTestInitializer.java [hunk 3]
```diff
@@ -100,6 +100,7 @@ public void truncateTables() throws Exception {
         tables.add("alert_triggers_alert_notifiers");
         tables.add("alert_triggers");
         tables.add("alert_notifiers");
+        tables.add("mfa_codes");
```
Comment (tcompiegne) [inline gravitee-am-repository/gravitee-am-repository-jdbc/src/test/java/io/gravitee/am/repository/jdbc/common/JdbcRepositoriesTestInitializer.java:103]: To remove

File: gravitee-am-resource/gravitee-am-resource-smtp/src/main/resources/schemas/schema-form.json [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+      "description": "Does the SMTP server requires authentication",
+      "default": false
+    },
+    "username": {
+      "type": "string",
+      "title": "Username",
+      "description": "Username used for authentication",
+      "condition": "model.authentication"
+    },
+    "password": {
```
Comment (tcompiegne) [inline gravitee-am-resource/gravitee-am-resource-smtp/src/main/resources/schemas/schema-form.json:41]: Add "widget": "password" to hide the password value

File: gravitee-am-repository/gravitee-am-repository-jdbc/src/test/java/io/gravitee/am/repository/jdbc/common/JdbcRepositoriesTestInitializer.java (hunk 3 above)
Comment (leleueri) [reply gravitee-am-repository/gravitee-am-repository-jdbc/src/test/java/io/gravitee/am/repository/jdbc/common/JdbcRepositoriesTestInitializer.java:103]: done

File: gravitee-am-resource/gravitee-am-resource-smtp/src/main/resources/schemas/schema-form.json (hunk 4 above)
Comment (leleueri) [reply gravitee-am-resource/gravitee-am-resource-smtp/src/main/resources/schemas/schema-form.json:41]: done

File: gravitee-am-factor/gravitee-am-factor-email/src/main/java/io/gravitee/am/factor/email/utils/SharedSecret.java [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.gravitee.am.factor.email.utils;
+
+import com.google.common.io.BaseEncoding;
+
+import java.security.SecureRandom;
+
+public final class SharedSecret {
```
[the remark this answers is not in the document]
Comment (leleueri) [reply gravitee-am-factor/gravitee-am-factor-email/src/main/java/io/gravitee/am/factor/email/utils/SharedSecret.java:22]: Done, moved this class to the factor-reporter-api module and add an additional method

File: gravitee-am-factor/gravitee-am-factor-email/src/main/java/io/gravitee/am/factor/email/provider/EmailFactorProvider.java [hunk 6]
```diff
@@ -168,17 +143,45 @@ public Completable sendChallenge(FactorContext context) {
...
-            return provider.sendMessage(new Message(recipient, content, subject, configuration.getContentType()));
+            final String recipient = enrolledFactor.getChannel().getTarget();
+
+            io.gravitee.am.common.email.Email email = new EmailBuilder()
+                    .from(mfaTemplate.getFrom())
+                    .fromName(mfaTemplate.getFromName())
+                    .subject(subject)
+                    .to(recipient)
+                    .build();
+            email.setContent(content);
```
Comment (tcompiegne) [inline gravitee-am-factor/gravitee-am-factor-email/src/main/java/io/gravitee/am/factor/email/provider/EmailFactorProvider.java]: I wonder if we cannot do all this technical stuff via the emailService ... something like : 

`Single<Email> emailService.createEmailTemplate(template, params, client);`

This way we remove all our internal templating mechanism (freemarker, ...)

**Pull request merged**